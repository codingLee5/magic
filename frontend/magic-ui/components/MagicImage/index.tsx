import type { ImgHTMLAttributes } from "react"
import { memo, useEffect, useState } from "react"

export interface MagicImageProps extends ImgHTMLAttributes<HTMLImageElement> {
	errorSrc?: string
}

const MagicImage = memo(function MagicImage(props: MagicImageProps) {
	const { errorSrc, ...imgProps } = props

	const [src, setSrc] = useState<string>(imgProps.src || "")

	useEffect(() => {
		if (imgProps.src) {
			setSrc(imgProps.src)
		}
	}, [imgProps.src])

	return (
		<img
			{...imgProps}
			src={src || errorSrc}
			onError={(event) => {
				setSrc(errorSrc || "")
				imgProps.onError?.(event)
			}}
		/>
	)
})

export default MagicImage
