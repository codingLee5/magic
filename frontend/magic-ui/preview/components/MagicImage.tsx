import React from "react"
import { Space } from "antd"
import MagicImage from "../../components/MagicImage"
import ComponentDemo from "./Container"

const ImageDemo: React.FC = () => {
	return (
		<div>
			<ComponentDemo
				title="基础用法"
				description="支持所有原生 img 标签的属性，使用方式与普通 img 标签一致"
				code="src: string"
			>
				<Space wrap>
					<MagicImage
						src="https://picsum.photos/200/150?random=1"
						alt="随机图片1"
						width={200}
						height={150}
						style={{ borderRadius: "8px" }}
					/>
					<MagicImage
						src="https://picsum.photos/200/150?random=2"
						alt="随机图片2"
						width={200}
						height={150}
						style={{ borderRadius: "8px" }}
					/>
					<MagicImage
						src="https://picsum.photos/200/150?random=3"
						alt="随机图片3"
						width={200}
						height={150}
						style={{ borderRadius: "8px" }}
					/>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="错误回退"
				description="当图片加载失败时，会自动显示 errorSrc 指定的图片"
				code="errorSrc: string"
			>
				<Space wrap>
					<MagicImage
						src="invalid-url"
						errorSrc="https://picsum.photos/200/150?random=fallback"
						alt="错误回退示例"
						width={200}
						height={150}
						style={{ borderRadius: "8px" }}
					/>
					<MagicImage
						src="https://picsum.photos/200/150?random=4"
						errorSrc="https://picsum.photos/200/150?random=fallback2"
						alt="正常图片"
						width={200}
						height={150}
						style={{ borderRadius: "8px" }}
					/>
				</Space>
			</ComponentDemo>
		</div>
	)
}

export default ImageDemo
