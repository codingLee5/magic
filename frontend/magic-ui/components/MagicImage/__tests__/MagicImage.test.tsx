import { render, screen } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicImage from "../index"

const renderWithTheme = (component: React.ReactElement) =>
	render(<MagicThemeProvider theme="light">{component}</MagicThemeProvider>)

describe("MagicImage", () => {
	it("应该正常渲染", () => {
		renderWithTheme(<MagicImage src="test.jpg" alt="测试图片" />)
		expect(screen.getByAltText("测试图片")).toBeInTheDocument()
	})

	// 快照测试
	describe("快照测试", () => {
		it("基础图片快照", () => {
			const { asFragment } = renderWithTheme(<MagicImage src="test.jpg" alt="测试图片" />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带类名图片快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicImage src="test.jpg" alt="测试图片" className="custom-image" />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带点击事件图片快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicImage src="test.jpg" alt="测试图片" onClick={() => {}} />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("自定义宽度图片快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicImage src="test.jpg" alt="测试图片" width={200} />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("自定义高度图片快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicImage src="test.jpg" alt="测试图片" height={150} />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带样式图片快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicImage src="test.jpg" alt="测试图片" style={{ borderRadius: "8px" }} />,
			)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
