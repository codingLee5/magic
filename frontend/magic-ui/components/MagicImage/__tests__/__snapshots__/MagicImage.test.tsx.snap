// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicImage > 快照测试 > 基础图片快照 1`] = `
<DocumentFragment>
  <img
    alt="测试图片"
    src="test.jpg"
  />
</DocumentFragment>
`;

exports[`MagicImage > 快照测试 > 带样式图片快照 1`] = `
<DocumentFragment>
  <img
    alt="测试图片"
    src="test.jpg"
    style="border-radius: 8px;"
  />
</DocumentFragment>
`;

exports[`MagicImage > 快照测试 > 带点击事件图片快照 1`] = `
<DocumentFragment>
  <img
    alt="测试图片"
    src="test.jpg"
  />
</DocumentFragment>
`;

exports[`MagicImage > 快照测试 > 带类名图片快照 1`] = `
<DocumentFragment>
  <img
    alt="测试图片"
    class="custom-image"
    src="test.jpg"
  />
</DocumentFragment>
`;

exports[`MagicImage > 快照测试 > 自定义宽度图片快照 1`] = `
<DocumentFragment>
  <img
    alt="测试图片"
    src="test.jpg"
    width="200"
  />
</DocumentFragment>
`;

exports[`MagicImage > 快照测试 > 自定义高度图片快照 1`] = `
<DocumentFragment>
  <img
    alt="测试图片"
    height="150"
    src="test.jpg"
  />
</DocumentFragment>
`;
